design/one-pager-composition-environment-beta.md

```diff
@@ -0,0 +1,381 @@
+clarity around the patch application order.
+
+We could define a few stages where we could in some way make it easier for
+users to gain visibility into the `in-memory environment`, for example:
+
+- resulting from the merge between the default data and all selected
+  `EnvironmentConfig`s
+- after having applied the patches at `spec.environment.patches`
+- after having applied all patches, before throwing it away
+
+See [here](https://github.com/crossplane/crossplane/pull/4702)  and
+[here](https://github.com/crossplane/crossplane/issues/3967) for some related
+discussions and possible approaches.
+
+On the other hand, beta composition functions have already improved this aspect
+by allowing the very same logic to be run locally using `crossplane beta
+render`, possibly running against deployed functions in the near future, and
+enabling the injection of arbitrary code at any point in the pipeline through
+the usage of a dedicated function, e.g. `function-debug`. Allowing Composition
+authors to do "print debugging" as in many other languages. So, we could
+postpone any action on this and rely on the improved experience enabled by
+Composition Functions, given we are
+[proposing](https://github.com/crossplane/crossplane/issues/4746) to deprecate
+the native P&T Compositions in the near future.
+
+## Next steps
+
+We currently believe https://github.com/crossplane/crossplane/issues/4739 would
+solve all the above issues:
+
+- Independent Managed Resource referencing:
+    - functions could request any Crossplane resource, so EnvironmentConfigs when it make sense to, or resources directly when it doesn't.
+- Naming Things:
+    - the concept of `in-memory environment` would just a convention between functions, relying on the [`apiextensions.crossplane.io/environment`](http://apiextensions.crossplane.io/environment) key in the context, as `function-patch-and-transform` already does.
+    - `--enable-environment-configs` would really mean only enabling `EnvironmentConfigs` as the rest of the `Composition Environment` would be taken care of by external functions, making the split between the two concepts clearer.
+- Selecting and merging EnvironmentConfigs:
+    - selection would be done by a dedicated function, e.g. `function-select-extra-resources`.
+    - merging into the environment would be done by another dedicated function, e.g. `function-set-environment` or a more generic `function-set-context`.
+    - **N.B.**: how to preserve resolve policy `IfNotPresent` behaviour? could be left to the selection function to implement by adding a place in the composite resource status to store referenced resources.
+- Debugging:
+    - visibility already addressed in the dedicated section above
+    - selection would be more visible too by allowing passing resources via `--extra-resources` to `crossplane beta render`.
+
+So we would have to:
+
+- Focus on https://github.com/crossplane/crossplane/issues/4739 for 1.15
```

Comment:
Just catching up with the latest state there. I need some time to process but my early thoughts are:

I am not yet convinced that the approach there should be the way we support composition environments while it might make sense for requesting extra resources. I feel like we need a more "native" support for environments rather than supporting them as a side effect (of capability of requesting extra resources) with a good amount of boilerplate code and 1-2 function calls each responsible with some part. This is mostly based on the fact that we know composition environments are crucial for any non-trivial composition and the desire to provide a smooth UX to enable this. So, my mind still goes to a solution like [this](https://github.com/crossplane/crossplane/issues/4583#issuecomment-1732343924) instead of splitting this functionality to dedicated functions.

If I try to make an analogy:

We are designing UNIX and discussing how someone should pass some variables into commands and instead of defining/inventing the thing we know as Environment Variables in UNIX, we are talking about suggesting something like the following as a solution for those who need such a functionality:

```shell
env-requester | env-merger | cmd1 | cmd 2
```

instead of:

```shell
export VAR1=value1
source /path/to/somefile
cmd1 | cmd2
```

I know this is not the perfect example, so please don't list the differences between a UNIX environment and K8s/Crossplane etc, the point I am trying to make is more about embracing a widely needed functionality and providing the best possible UX with native support vs "enabling" it somehow.

Reply:
As per your example, the two things don't exclude each other, in shell scripts, piped commands have access to the environment.

To continue with the UNIX analogy, I see enabling functions to retrieve resources more like providing UNIX sockets and leaving to users to decide what to do with them. 

What I'm suggesting is indeed, given that we have a finite amount of time, to put on hold the development of a native approach and instead focus on enabling functions. Nothing would block us from "bringing back in" as native the "best possible UX" once it emerges from functions and we agree on it.